filesys/handler.go

```diff
@@ -0,0 +1,300 @@
+		fa.handle(&Close{}, w, r)
+	case "/fs/rename":
+		fa.handle(&Rename{}, w, r)
+	case "/fs/readdir":
+		fa.handle(&Readdir{}, w, r)
+	case "/fs/lstat":
+		fa.handle(&Lstat{}, w, r)
+	case "/fs/read":
+		fa.handle(&Read{}, w, r)
+	case "/fs/mkdir":
+		fa.handle(&Mkdir{}, w, r)
+	case "/fs/unlink":
+		fa.handle(&Unlink{}, w, r)
+	case "/fs/rmdir":
+		fa.handle(&Rmdir{}, w, r)
+	default:
+		fa.doError("not implemented", "ENOSYS", w)
+	}
+}
+
+type Responder interface {
+	WriteResponse(fa *Handler, w http.ResponseWriter)
+}
+
+func (fa *Handler) handle(responder Responder, w http.ResponseWriter, r *http.Request) {
+	if err := json.NewDecoder(r.Body).Decode(responder); err != nil {
+		fa.logger.Printf("ERROR handle : %s\n", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if fa.debug {
+		fa.logger.Printf("handle %s %+v\n", r.URL.Path, responder)
+	}
+	responder.WriteResponse(fa, w)
+}
+
+type ErrorCode struct {
+	Error string `json:"error"`
+	Code  string `json:"code"`
+}
+
+func (fa *Handler) doError(msg, code string, w http.ResponseWriter) {
+	if fa.debug {
+		fa.logger.Printf("doError %s : %s\n", msg, code)
+	}
+	e := &ErrorCode{Error: msg, Code: code}
+
+	w.WriteHeader(http.StatusBadRequest)
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(e); err != nil {
+		fa.logger.Println("doError json error :", err)
+	}
+}
+
+func (fa *Handler) okResponse(data any, w http.ResponseWriter) {
+	if marshal, err := json.Marshal(data); err != nil {
+		fa.logger.Println("okResponse json error:", err)
+		w.WriteHeader(http.StatusInternalServerError)
+	} else {
+		if fa.debug {
+			fa.logger.Printf("okResponse %s\n", string(marshal))
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write(marshal)
+	}
+}
+
+func fixPath(path string) string {
+	return strings.TrimPrefix(path, "/fs/")
+}
+
+type Stat struct {
+	Path string `json:"path,omitempty"`
+}
+
+type Open struct {
+	Path  string `json:"path"`
+	Flags int    `json:"flags"`
+	Mode  uint32 `json:"mode"`
+}
+
+func (o *Open) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	fd, err := syscall.Open(fixPath(o.Path), o.Flags, o.Mode)
+	if fa.handleError(w, err, true) {
+		return
+	}
+	response := map[string]any{"fd": fd}
+	fa.okResponse(response, w)
+}
+
+type Fstat struct {
+	Fd int `json:"fd"`
+}
+
+type Write struct {
+	Fd       int    `json:"fd"`
+	Buffer   string `json:"buffer"`
+	Offset   int    `json:"offset"`
+	Length   int    `json:"length"`
+	Position *int   `json:"position,omitempty"`
+}
+
+func (wr *Write) WriteResponse(fa *Handler, w http.ResponseWriter) {
+
+	if wr.Position != nil || wr.Offset != 0 {
+		fa.doError("not implemented", "ENOSYS", w)
+		return
+	}
+
+	bytes, err := base64.StdEncoding.DecodeString(wr.Buffer)
+	if err != nil {
+		fa.doError("not implemented", "ENOSYS", w)
+		return
+	}
+
+	var written int
+	written, err = syscall.Write(FdType(wr.Fd), bytes)
+	if err != nil {
```

Comment:
It's not exactly clear to me. Sometimes we are calling `fa.handleError` and sometimes we are doing the error handling manually. Is that intentional or just some refactoring is needed here?

Reply:
The `fa.handleError` usage was to reduce code in methods where a non nil error needs to be ENOENT or ENOSYS based on os.IsNotExist() . It also uses syscall.ENOSYS.Error() instead of "not implemented" for the message which I believe is checked by go code handling the error.